proposals/route-blinding.md

```diff
@@ -0,0 +1,613 @@
+
+More generally, aggregated fees can be computed recursively with the following formulas, replacing `ceil(a/b)` with `(a+b-1)/b`:
+
+* `route_fee_base_msat(n+1) = (fee_base_msat(n+1) * 1000000 + route_fee_base_msat(n) * (1000000 + fee_proportional_millionths(n+1)) + 1000000 - 1) / 1000000`
+* `route_fee_proportional_millionths(n+1) = ((route_fee_proportional_millionths(n) + fee_proportional_millionths(n+1)) * 1000000 + route_fee_proportional_millionths(n) * fee_proportional_millionths(n+1) + 1000000 - 1) / 1000000`
+
+Let's assume the current block height is 1000. Alice wants the route to be used in the next 200
+blocks, so she sets `max_cltv_expiry = 1200` and adds `cltv_expiry_delta` for each hop. Alice then
+transmits the following information to the sender (most likely via an invoice):
+
+* Blinded route: `[N(carol), B(bob), B(alice)]`
+* First blinding ephemeral key: `E(carol)`
+* Aggregated route relay parameters and constraints:
+  * `fee_base_msat`: 201
+  * `fee_proportional_millionths`: 1001
+  * `htlc_minimum_msat`: 1000
+  * `cltv_expiry_delta`: 288
+  * `max_cltv_expiry`: 1200
+  * `allowed_features`: empty
+* Encrypted data for blinded nodes:
+  * `encrypted_payload(alice)`:
+    * `path_id`: `payment_preimage`
+    * `max_cltv_expiry`: 1200
+  * `encrypted_payload(bob)`:
+    * `outgoing_channel_id`: `scid_bob_alice`
+    * `fee_base_msat`: 100
+    * `fee_proportional_millionths`: 500
+    * `htlc_minimum_msat`: 1000
+    * `max_cltv_expiry`: 1344
+  * `encrypted_payload(carol)`:
+    * `outgoing_channel_id`: `scid_carol_bob`
+    * `fee_base_msat`: 100
+    * `fee_proportional_millionths`: 500
+    * `htlc_minimum_msat`: 1000
+    * `max_cltv_expiry`: 1488
+
+Note that the introduction point (Carol) uses the real `node_id`, not the blinded one, because the
+sender needs to be able to locate this introduction point and find a route to it. The sender will
+send the first blinding ephemeral key `E(carol)` in the onion `hop_payload` for Carol, which will
+allow Carol to compute the blinding shared secret and correctly forward. We put this blinding
+ephemeral key in the onion instead of using a tlv in `update_add_htlc` because intermediate nodes
+added before the blinded route may not support route blinding and wouldn't know how to relay it.
+
+Erin wants to send 100 000 msat to this blinded route.
+She can reach Carol via Dave: `Erin -> Dave -> Carol`, where the channel between Dave and Carol uses
+the following relay parameters:
+
+* `fee_base_msat`: 10
+* `fee_proportional_millionths`: 100
+* `cltv_expiry_delta`: 24
+
+Erin uses the aggregated route relay parameters to compute how much should be sent to Carol:
+
+* `amount = ceil(100000 + 201 + 1001 * 100000 / 1000000) = 100302 msat`
+
+Erin chooses a final expiry of 1100, which is below Alice's `max_cltv_expiry`, and computes the
+expiry that should be sent to Carol:
+
+* `expiry = 1100 + 288 = 1388`
```

Comment:
> Why does the recipient/route blinder not communicate to the sender (Erin) the exact min_final_cltv_expiry she should use

It's not about setting a *min* expiry, it's the opposite, this is about setting a *max* expiry, which is a concept that doesn't exist with Bolt 11 invoices. We want to make the route expire after a while because:

- we expect to receive that payment soon, so there's no reason to allow using the blinded route for too long
- the longer a blinded route can be used for payments, the higher the risk is that an attacker may be able to probe it and potentially unblind some of the nodes inside it

This is completely independent of the `cltv_expiry_delta` / `min_final_cltv_expiry` mechanisms.

Reply:
> It's not about setting a min expiry, it's the opposite, this is about setting a max expiry, which is a concept that doesn't exist with Bolt 11 invoices. We want to make the route expire after a while because...

That part comes through loud and clear. We're certainly defining a window in which payments to the blinded route will have a chance to succeed. I was thinking that the *`height + min_final_cltv_expiry`* forms the "left" boundary of that window while the *`max_cltv_expiry`* forms the right side. 

I was missing a critical piece: The *`min_final_cltv_expiry`* is not a constraint specified in channel/edge policy or learned via gossip. It's an invoice level parameter, so it can be shared with the sender along side a blinded route via an invoice. That's what I get for not knowing much about invoice structure 😅 Thanks for the help in getting that straight.